sycl/include/CL/sycl/handler.hpp

```diff
@@ -729,6 +810,305 @@ class __SYCL_EXPORT handler {
+        NWorkGroups == 1;
+
+    // Use local memory to reduce elements in work-groups into 0-th element.
+    // If WGSize is not power of two, then WGSize+1 elements are allocated.
+    // The additional last element is used to catch reduce elements that
+    // could otherwise be lost in the tree-reduction algorithm.
+    auto LocalReds = Redu.getReadWriteLocalAcc(WGSize + InefficientCase, *this);
+
+    // Get read accessor to the buffer that was used as output
+    // in the previous kernel. After that create new output buffer if needed
+    // and get accessor to it (or use reduction's accessor if the kernel
+    // is the last one).
+    auto In = Redu.getReadAccToPreviousPartialReds(*this);
+    auto Out = Redu.getWriteAccForPartialReds(NWorkGroups, KernelRun, *this);
+
+    if (!InefficientCase) {
+      // Efficient case: work-groups are fully loaded and work-group size
+      // is power of two.
+      using AuxName = typename detail::get_reduction_aux_1st_kernel_name_t<
+          KernelName, KernelType>::name;
+      parallel_for<AuxName>(Range, [=](nd_item<Dims> NDIt) {
+        // Copy the element to local memory to prepare it for tree-reduction.
+        size_t LID = NDIt.get_local_linear_id();
+        size_t GID = NDIt.get_global_linear_id();
+        LocalReds[LID] = In[GID];
+        NDIt.barrier();
+
+        // Tree-reduction: reduce the local array LocalReds[:] to LocalReds[0]
+        typename Reduction::binary_operation BOp;
+        size_t WGSize = NDIt.get_local_range().size();
+        for (size_t CurStep = WGSize >> 1; CurStep > 0; CurStep >>= 1) {
+          if (LID < CurStep)
+            LocalReds[LID] = BOp(LocalReds[LID], LocalReds[LID + CurStep]);
+          NDIt.barrier();
+        }
+
+        // Compute the the partial sum/reduction for the work-group.
+        if (LID == 0)
+          Out.get_pointer().get()[NDIt.get_group_linear_id()] =
+              IsUpdateOfUserAcc ? BOp(*(Out.get_pointer()), LocalReds[0])
+                                : LocalReds[0];
+      });
+    } else {
+      // Inefficient case: work-groups are not fully loaded
+      // or WGSize is not power of two.
+      // These two inefficient cases are handled by one kernel, which
+      // can be split later into two separate kernels, if there are users
+      // who really need more efficient code for them.
+      using AuxName = typename detail::get_reduction_aux_2nd_kernel_name_t<
+          KernelName, KernelType>::name;
+      auto ReduIdentity = Redu.getIdentity();
+      parallel_for<AuxName>(Range, [=](nd_item<Dims> NDIt) {
+        size_t WGSize = NDIt.get_local_range().size();
+        size_t LID = NDIt.get_local_linear_id();
+        size_t GID = NDIt.get_global_linear_id();
+        // Copy the element to local memory to prepare it for tree-reduction
+        LocalReds[LID] = (GID < NWorkItems) ? In[GID] : ReduIdentity;
+        LocalReds[WGSize] = ReduIdentity;
```

Comment:
Will it be better to allow only one `NDIt` write to `WGSize` to not have multiple write operations from different items?

Reply:
Excuse me, I do not understand what you suggest here. Please give more details.

Reply:
I mean that you have `WGSize` work items: `0,1,..., WGSize-1`. Each of work-items updates `LocalReds[WGSize]`. Would it better to have something like that `if(LID==0){LocalReds[WGSize] = ...}`. So, only one work item (not all) updates the value.

Reply:
If you don't mind I'll try your suggestion in the next patch.
It is a tiny fix that should not hold the commit, because it will not change performance in any way.

Having additional conditional in the code is also bad for performance. (Hmm, both the conditional and that duplicated write could be eliminated by vectorizer that combines ops across work-items), 
I'll check the device code and do or not do the additional fix with the next patch which is currently blocked by this PR.

Reply:
OK. Thanks for the response